File: c/linkedlist.c
/* This is singly linked list implementation program.
Only head and right(next) pointer is used*/

#include<stdio.h>
#include<stdlib.h>
#include "list.h"

int main(){
  char ch;
  int num,key;
  lnodeptr ptr;
  head = NULL;/*initializing head ptr*/
  printf("Insert(i)/Search(s)/Delete(d)/Exit(e) : ");
  while((ch=getchar()) != 'e'){
    switch(ch){
    case 'i':
      printf("Enter no. of nodes:");
      scanf("%d",&num);
      for(;num>0;num--){
        printf("Enter key:");
        scanf("%d",&key);
        if(head != NULL)
          insertNode(createNode(key));
        else
          head = createNode(key);
      }
      printf("Head : %d,\tList : ",head->key);
      printList(head);
      printf("\n");
      break;
    case 's':
      printf("Key value to be searched:");
      scanf("%d",&key);
      if((ptr=searchNode(key)) != NULL)
        printf("Found : %d\t Stored at : %p\n",ptr->key,ptr);
        else
          printf("Oopss !! No such key found\n");
      break;
    case 'd':
      printf("Key value to be deleted:");
      scanf("%d",&key);
      if((ptr=searchNode(key)) != NULL){
        key=deleteNode(ptr);
        printf("Node deleted\n");
        printf("Head : %d\nList : ",head->key);
        printList(head);
        //printf("\n");
      }
      else
        printf("Oopss !! No such key exists\n");
      break;
    default:
      printf("Wrong choice, try again\n");
      break;
    }
    while((ch=getchar()) != '\n' && ch != EOF);
    printf("Insert(i)/Search(s)/Delete(d)/Exit(e) : ");
  }
  return 0;
}

File: go/src/linkedlist.go
package main

import (
    "fmt"
    //"os"
)

type lnode struct{
    key int
    next *lnode
}

func insert(head *lnode, val int) *lnode{
    temp := new(lnode)
    temp.key = val
    temp.next = head.next
    head.next = temp
    return head
}

func delete(head *lnode, val int){
    curr := head
    for curr.next != nil{
        if curr.next.key == val{
            curr.next = curr.next.next
            return
            }
        curr = curr.next
    }
}

func printlist(head *lnode){
    curr := head.next
    for curr != nil{
        fmt.Printf("%d->", curr.key)
        curr = curr.next
    }
    fmt.Println(curr)
}

func main(){
    head := new(lnode)
    for i:= 1; i < 10; i++{
        head = insert(head, i)
    }
    printlist(head)
    delete(head, 1)
    printlist(head)
    delete(head, 5)
    printlist(head)
    delete(head, 9)
    printlist(head)
}
